app/lib/analyzer/backend.dart

```diff
@@ -203,27 +203,50 @@ class AnalysisBackend {
     return true;
   }
 
-  Future deleteObsoleteAnalysisEntries() async {
+  /// Deletes the obsolete [Analysis] instances from Datastore. An instance is
+  /// obsolete, if:
+  /// - it is **not** the latest for the pana version, or
+  /// - it is older than 6 months (except if it is the latest one).
+  Future deleteObsoleteAnalysis(String package, String version) async {
     final DateTime threshold =
         new DateTime.now().toUtc().subtract(obsoleteThreshold);
-    final Query query = db.query(Analysis)
-      ..order('timestamp')
-      ..filter('timestamp <', threshold);
-    await for (Analysis analysis in query.run()) {
-      final PackageVersionAnalysis pva =
-          (await db.lookup([analysis.parentKey])).single;
-      if (pva.latestAnalysis != analysis.analysis) {
-        _logger.info('Deleting obsolete Analysis: ${analysis.packageName} '
-            '${analysis.packageVersion} ${analysis.analysis}');
-        await db.withTransaction((Transaction tx) async {
-          final List list = await tx.lookup([analysis.key]);
-          if (list[0] != null) {
-            tx.queueMutations(deletes: [list[0].key]);
-          }
-          await tx.commit();
-        });
+    final Query scanQuery = db.query(
+      Analysis,
+      ancestorKey: db.emptyKey
+          .append(PackageAnalysis, id: package)
+          .append(PackageVersionAnalysis, id: version),
+    );
+    final List<Key> obsoleteKeys = <Key>[];
+
+    final List<_AnalysisRef> refs = <_AnalysisRef>[];
+    await for (Analysis analysis in scanQuery.run()) {
+      refs.add(new _AnalysisRef(
+          analysis.key, analysis.panaVersion, analysis.timestamp));
```

Comment:
No need for the `_AnalysisRef` class, all the information is available in `Analysis`, just use:

```dart
final List<Analysis> existingAnalysis = scanQuery.run().toList();
existingAnalysis.sort(...);
```